File: srcs/server/agents/builtin/budget.go
package builtin

import "fmt"

// BudgetTracker mimics the tracking logic in claude-code query/tokenBudget.ts.
type BudgetTracker struct {
	ContinuationCount    int
	LastDeltaTokens      int
	LastGlobalTurnTokens int
}

const (
	completionThreshold float64 = 0.9
	diminishingThreshold int = 500
)

type TokenBudgetDecision struct {
	Action            string // "continue" or "stop"
	NudgeMessage      string
	ContinuationCount int
	Pct               int
	TurnTokens        int
	Budget            int
	Diminishing       bool
}

// CheckTokenBudget evaluates if the model has hit a budget constraint.
// It is designed to be called when the LLM stops generating due to length
// (e.g. hit max_tokens for a turn) and we want to nudge it to continue
// towards the MaxTaskBudget.
func CheckTokenBudget(tracker *BudgetTracker, budget int, globalTurnTokens int) TokenBudgetDecision {
	if budget <= 0 {
		return TokenBudgetDecision{Action: "stop"}
	}

	pct := int((float64(globalTurnTokens) / float64(budget)) * 100)
	deltaSinceLastCheck := globalTurnTokens - tracker.LastGlobalTurnTokens

	isDiminishing := tracker.ContinuationCount >= 3 &&
		deltaSinceLastCheck < diminishingThreshold &&
		tracker.LastDeltaTokens < diminishingThreshold

	if !isDiminishing && float64(globalTurnTokens) < float64(budget)*completionThreshold {
		tracker.ContinuationCount++
		tracker.LastDeltaTokens = deltaSinceLastCheck
		tracker.LastGlobalTurnTokens = globalTurnTokens
		nudge := fmt.Sprintf("Stopped at %d%% of token target (%d / %d). Keep working — do not summarize.", pct, globalTurnTokens, budget)
		return TokenBudgetDecision{
			Action:            "continue",
			NudgeMessage:      nudge,
			ContinuationCount: tracker.ContinuationCount,
			Pct:               pct,
			TurnTokens:        globalTurnTokens,
			Budget:            budget,
		}
	}

	return TokenBudgetDecision{
		Action:            "stop",
		ContinuationCount: tracker.ContinuationCount,
		Pct:               pct,
		TurnTokens:        globalTurnTokens,
		Budget:            budget,
		Diminishing:       isDiminishing,
	}
}

File: src/agents/builtin/budget.rs
/// Token budget tracker — mirrors Go BudgetTracker.
#[derive(Debug, Default)]
pub struct BudgetTracker {
    pub continuation_count: i32,
    pub last_delta_tokens: i32,
    pub last_global_turn_tokens: i32,
}

const COMPLETION_THRESHOLD: f64 = 0.9;
const DIMINISHING_THRESHOLD: i32 = 500;

pub struct TokenBudgetDecision {
    pub action: BudgetAction,
    pub nudge_message: String,
    pub continuation_count: i32,
    pub pct: i32,
    pub turn_tokens: i32,
    pub budget: i32,
    pub diminishing: bool,
}

#[derive(Debug, PartialEq, Eq)]
pub enum BudgetAction {
    Continue,
    Stop,
}

/// Evaluate token budget state when the LLM stops due to length.
/// Mirrors Go's CheckTokenBudget.
pub fn check_token_budget(
    tracker: &mut BudgetTracker,
    budget: i32,
    global_turn_tokens: i32,
) -> TokenBudgetDecision {
    if budget <= 0 {
        return TokenBudgetDecision {
            action: BudgetAction::Stop,
            nudge_message: String::new(),
            continuation_count: tracker.continuation_count,
            pct: 0,
            turn_tokens: global_turn_tokens,
            budget,
            diminishing: false,
        };
    }

    let pct = ((global_turn_tokens as f64 / budget as f64) * 100.0) as i32;
    let delta = global_turn_tokens - tracker.last_global_turn_tokens;

    let is_diminishing = tracker.continuation_count >= 3
        && delta < DIMINISHING_THRESHOLD
        && tracker.last_delta_tokens < DIMINISHING_THRESHOLD;

    if !is_diminishing && (global_turn_tokens as f64) < (budget as f64) * COMPLETION_THRESHOLD {
        tracker.continuation_count += 1;
        tracker.last_delta_tokens = delta;
        tracker.last_global_turn_tokens = global_turn_tokens;
        let nudge = format!(
            "Stopped at {}% of token target ({} / {}). Keep working — do not summarize.",
            pct, global_turn_tokens, budget
        );
        return TokenBudgetDecision {
            action: BudgetAction::Continue,
            nudge_message: nudge,
            continuation_count: tracker.continuation_count,
            pct,
            turn_tokens: global_turn_tokens,
            budget,
            diminishing: false,
        };
    }

    TokenBudgetDecision {
        action: BudgetAction::Stop,
        nudge_message: String::new(),
        continuation_count: tracker.continuation_count,
        pct,
        turn_tokens: global_turn_tokens,
        budget,
        diminishing: is_diminishing,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_budget_continue_below_threshold() {
        let mut tracker = BudgetTracker::default();
        let decision = check_token_budget(&mut tracker, 1000, 400);
        assert_eq!(decision.action, BudgetAction::Continue);
        assert!(decision.nudge_message.contains("40%"));
    }

    #[test]
    fn test_budget_stop_above_threshold() {
        let mut tracker = BudgetTracker::default();
        let decision = check_token_budget(&mut tracker, 1000, 950);
        assert_eq!(decision.action, BudgetAction::Stop);
    }

    #[test]
    fn test_budget_zero() {
        let mut tracker = BudgetTracker::default();
        let decision = check_token_budget(&mut tracker, 0, 100);
        assert_eq!(decision.action, BudgetAction::Stop);
    }
    #[test]
    fn test_budget_diminishing_returns() {
        let mut tracker = BudgetTracker {
            continuation_count: 3,
            last_delta_tokens: 400,
            last_global_turn_tokens: 0,
        };
        let decision = check_token_budget(&mut tracker, 10000, 400); // delta is 400
        assert_eq!(decision.action, BudgetAction::Stop);
        assert!(decision.diminishing);
    }

}
